pkg/controller/controller.go

```diff
@@ -1154,37 +1183,56 @@ func (p *csiProvisioner) Delete(ctx context.Context, volume *v1.PersistentVolume
 	req := csi.DeleteVolumeRequest{
 		VolumeId: volumeId,
 	}
-	// get secrets if StorageClass specifies it
-	storageClassName := util.GetPersistentVolumeClass(volume)
-	if len(storageClassName) != 0 {
-		if storageClass, err := p.scLister.Get(storageClassName); err == nil {
-			if migratedVolume && storageClass.Provisioner == p.supportsMigrationFromInTreePluginName {
-				klog.V(2).Infof("translating storage class for in-tree plugin %s to CSI", storageClass.Provisioner)
-				storageClass, err = p.translator.TranslateInTreeStorageClassToCSI(p.supportsMigrationFromInTreePluginName, storageClass)
-				if err != nil {
-					return err
-				}
-			}
-
-			// Resolve provision secret credentials.
-			provisionerSecretRef, err := getSecretReference(provisionerSecretParams, storageClass.Parameters, volume.Name, &v1.PersistentVolumeClaim{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      volume.Spec.ClaimRef.Name,
-					Namespace: volume.Spec.ClaimRef.Namespace,
-				},
-			})
-			if err != nil {
-				return fmt.Errorf("failed to get secretreference for volume %s: %v", volume.Name, err)
-			}
 
+	if metav1.HasAnnotation(volume.ObjectMeta, AnnDeletionProvisionerSecretRefName) && metav1.HasAnnotation(volume.ObjectMeta, AnnDeletionProvisionerSecretRefNamespace) {
```

Comment:
This prefers annotations over StorageClass even when the StorageClass still exists. Is this intended behavior? What if the StorageClass was re-created with updated secrets?

Reply:
I think that other `Secret`s are stored in `csiSource` on [provision](https://github.com/kubernetes-csi/external-provisioner/blob/32c27a57ce311cf7b0fedefd6cc850845a41e341/pkg/controller/controller.go#L672-L679) and won't be changed when `StorageClass` is updated, so the same behavior, prioritizing the value on provision, would be also expected for provisioner-secret.

Although, we might still be able to discuss if the current behavior of the other `Secret`s is good or not.

Reply:
yeah, considering we give preference to the spec source once PV is bound, I believe SC recreation should also get less preference here.

Reply:
ack